pytorch_lightning/trainer/training_tricks.py

```diff
@@ -80,3 +85,196 @@ def configure_accumulated_gradients(self, accumulate_grad_batches):
+
+    def scale_batch_size(self,
+                         model: LightningModule,
+                         mode: str = 'power',
+                         n_step_per_try: int = 3,
+                         init_val: int = 2,
+                         n_max_try: int = 15):
+        r""" Will iteratively try to find the largest batch size for a given model
+            that does not not give an out of memory (OOM) error
+
+        Args:
+            model: Model to fit.
+
+            mode: string setting the search mode. Either `power` or `binsearch`.
+                If mode is `power` we keep multiplying the batch size by 2, until
+                we get an OOM error. If mode is 'binsearch', we will initially
+                also keep multiplying by 2 and after encountering an OOM error
+                do a binary search between the last succeded batch size and the
+                batch size that failed.
+
+            n_step_per_try: number of steps to run with a given batch size.
+                Idealy 1 should be enough to test if a OOM error occurs,
+                however in practise a few is needed
+
+           init_val: initial batch size to do the search from
+
+           n_max_try: max number of increase/decreases in batch size done before
+               algorithm is terminated
+
+        """
+        if mode not in ['power', 'binsearch']:
+            raise ValueError('mode in method `scale_batch_size`'
+                             ' can only be `power` or `binsearch')
+
+        # Arguments we adjust during the batch size finder, save for restoring
+        trainer_arg = self.auto_scale_batch_size
+        max_steps = self.max_steps
+        weights_summary = self.weights_summary
+        logger = self.logger
+        callbacks = self.callbacks
+        checkpoint_callback = self.checkpoint_callback
+
+        self.auto_scale_batch_size = False  # prevent recursion
+        self.max_steps = n_step_per_try  # take few steps
+        self.weights_summary = None  # not needed before full run
+        self.logger = None  # not needed before full run
+        self.callbacks = []  # not needed before full run
+        self.checkpoint_callback = False  # required for saving
+        self.optimizers, self.schedulers = [], []  # required for saving
+        self.model = model  # required for saving
+
+        # Save initial model, that is loaded after batch size is found
+        save_path = os.path.join(self.default_root_dir, 'temp_model.ckpt')
+        self.save_checkpoint(str(save_path))
+
+        # Start increase/decrease batch size
+        new_size = _adjust_batch_size(self, trainer_arg, value=init_val)  # initially set to init_val
+        high = None
+        count = 0
+        while True:
+            garbage_collection_cuda()
+            self.global_step = 0  # reset after each try
+            try:
+                # Try fit
+                self.fit(model)
```

Comment:
can't you use the sanity checks for that?

Reply:
I dont think so. Running with `num_sanity_val_steps>0` would not give the full memory footprint, since gradient information is not accounted for. Calling `fit` can quantify this.